### ISSUE TITLE
strange noises in your samples && error when running inference.py

### ISSUE THREAD

**Issue by MorganCZY** [seungwonpark/melgan#30]
Your samples at epoch 3200 have strange noises at unvoiced segments, while there is no such phenomenon in samples at epoch 1600.
![noise](https://user-images.githubusercontent.com/33675916/68571456-0137cf00-049e-11ea-8c5a-4feeaef9dcc5.png)
Besides, when running inference.py, an error occurs, pointing to https://github.com/seungwonpark/melgan/blob/8af1e9c12e22f28186739664febb6b0285e3e212/model/generator.py#L68 torch.cat() has a parameter "dim" rather than "axis" 
![cat](https://user-images.githubusercontent.com/33675916/68571860-f7fb3200-049e-11ea-9dc3-b56cab71969b.png)

**Comment by seungwonpark** [comment]
- Fixed the latter issue, thank you!
- Yes, I was aware of that issue. I also found out that results from mel-spectrogram generated from 0-filled audio (which is `-11.5129`-filled mel) are very noisy. I really need to solve this issue.

**Comment by seungwonpark** [comment]
[zeromel.zip](https://github.com/seungwonpark/melgan/files/3830437/zeromel.zip)

Spectrogram of zero-filled audio reconstruction looks like this: the line noise appears every 4 frequency bins.

EDIT: total frequency bins are 512. So the pattern appears for every 4 bins, not 8. The y-axis of the figure below is wrong.

![zeromel](https://user-images.githubusercontent.com/13394294/68575320-23861880-04af-11ea-8c0f-8cddead736d0.png)

**Comment by seungwonpark** [comment]
I hope to fix it by matching the implementation details with official implementation. See #17.

**Comment by MorganCZY** [comment]
I have trained and tested the official MelGAN repo. Synthesized samples are with audible noises. The overall effect is far less than the official pretrained model.

**Comment by seungwonpark** [comment]
Oh, does that mean we need to use some tricks (that aren’t shown in paper) to properly train the model?

**Comment by MorganCZY** [comment]
I highly doubt there are some training tricks that are not shown in the official repo codes. I left an issue at their repo, but haven't received a reply till now.

**Comment by bob80333** [comment]
Checkboard artifacts have been an issue with image GANs before, see this article: https://distill.pub/2016/deconv-checkerboard/

I think some of these audio artifacts may be related.  The main way to get rid of them was to replace strided conv layers with bilinear upsample/downsample + conv layers, or to ensure that kernels were exact multiples of their strides.  The discriminator here appears to have kernels of 41 with strides of 4, I wonder what would happen if we stuck a bilinear downsample 4x before those convs, and set the stride to 1.

I'm going to try this out myself, but first I'm waiting for a model I'm training on part of VoxCeleb2 (the full dataset doesn't fit in my ssd) to hit 1M training steps before I try this to see if there's any improvement.

**Comment by seungwonpark** [comment]
Nice point, but isn't it a problem of a generator? The generator architecture doesn't seem to have that kind of problem, but only discriminator does.

**Comment by bob80333** [comment]
At the end of that article, just before the conclusion, they found that discriminators with stride=2 in the first layer could also cause the generator to create the checkboard artifacts.  The explanation was that some of the neurons in the generator will get many times the gradient due to the striding in the discriminator, and that helps create the artifacts.  

I don't know if that would apply to this audio GAN, but it seems like a fairly simple thing to check.  I have modified the discriminator in my fork, and I will start a training run tonight to see if it helps.

**Comment by bob80333** [comment]
Tested my fork out, the discriminator converges really fast, and the generator learns nothing.

<img width="424" src="https://user-images.githubusercontent.com/6364377/68952806-8b2db380-078e-11ea-9f2d-978059384d03.png">

Note the scales here
<img width="424"  src="https://user-images.githubusercontent.com/6364377/68952810-8d900d80-078e-11ea-8812-481d2c7e8060.png">

What the generator's output looks like:

<img width="396" alt="Screen Shot 2019-11-15 at 9 58 21 AM" src="https://user-images.githubusercontent.com/6364377/68952886-b1ebea00-078e-11ea-967f-3cc6be8dd642.png">
<img width="408" alt="Screen Shot 2019-11-15 at 9 58 13 AM" src="https://user-images.githubusercontent.com/6364377/68952889-b3b5ad80-078e-11ea-98bc-3f68a97241f2.png">

Swapping from strided convolutions to downsampling appears to have made the discriminator much stronger, not sure how to fix that...

**Comment by seungwonpark** [comment]
I feel sorry to hear that.

Is using nn.Upsample for downsampling is okay? The documentation says
> If you want downsampling/general resizing, you should use interpolate

Thanks for sharing your results, by the way.

**Comment by bob80333** [comment]
Oh!  Nice catch, I missed that in the docs.  I just fixed it in my fork, training is slightly better with this, but the discriminator still overpowers the generator quickly.

Discriminator converged in 2k steps rather than <500 steps.

<img width="365" alt="Screen Shot 2019-11-15 at 4 03 43 PM" src="https://user-images.githubusercontent.com/6364377/68975604-af56b800-07c1-11ea-94f2-582096deaae8.png">

**Comment by geekboood** [comment]
@bob80333 Hi, I try to train the melgan on csmsc dataset, which is a single speaker dataset about 20 hours. My understand is the discriminator should converge pretty fast because at the very beginning the generator's result is indeed very easy to discern, since the result is very bad. And if you run for more epoches, you may find that the generator's result is improved at some time. Here is my tensorboard log. 
![SharedScreenshot](https://user-images.githubusercontent.com/23272969/69001881-77668680-0921-11ea-8802-f9250b4c0a0c.jpg)
As you can see in the figure, the generator's loss stuck at around 120 before 300k iteration and after that, the loss is getting good. At the same time, the loss of the discriminator fluctuates a lot. I can hear something after 1.1M step, but it still with some artifacts. Maybe I should wait for 2M iteration. 
Also I found that at the end of each audio, there exist a peak that generates the noise.

**Comment by bob80333** [comment]
Hey, thanks for the information!  I have trained on my dataset (part of VoxCeleb2) with the current master branch for 1M steps and got this training curve:

<img width="406" alt="Screen Shot 2019-11-16 at 9 23 54 PM" src="https://user-images.githubusercontent.com/6364377/69002093-89abda80-08b7-11ea-847f-d559a7457c1c.png">

The results were understandable, but the voices themselves had artifacts while speaking, which is why I commented in this issue with ideas to fix it.  The first modification I tried, I waited 80k steps, at which point the discriminator had gotten to 3.3e-5 loss and the generator was generating loud high pitched noises.  I tried other approaches but the discriminator converged really quickly again, and I didn't want to wait to see if it failed, especially since my original training curve was very different from that.

**Comment by seungwonpark** [comment]
I've trained with `fix/17` branch for 14 days (more than 6400 epochs) with LJSpeech-1.1 dataset, and the results don't have strange noise at unvoiced segment! I'll soon upload new audio samples(with pre-trained model, if possible), and merge `fix/17` branch to `master`.

**Comment by seungwonpark** [comment]
Issues that were initially discussed here are now resolved, but I loved the idea and countless trials of @bob80333 to improve the quality.
Feel free to have more discussion here, or you may want to open a new issue.
